### ISSUE TITLE
Sources won't compile (linux)

### ISSUE THREAD

**Issue by rstaub77** [ethanpil/sheret#4]
Hi,
I tried to compile from sources (latest)
~~~~bash
cd build
go build -ldflags="-s -w" ../src/sheret.go
~~~~
All I got is
~~~~
../src/sheret.go:20:3: cannot find package "github.com/pkg/browser" in any of:
(list of local go dependencies cache)
~~~~
So there seem to be a missing dependance after all.
I tried to remove the dep from the include list, but the so-called "browser" package is effectively in use, and another error arises.
Any idea?

**Comment by marekfilip** [comment]
Hi,
I had the same problem, it's because sheret depends on it, you have to download them.

Here, I got your back, do as following:
1. `go get github.com/ethanpil/sheret`
You will receive message like `package github.com/ethanpil/sheret: no Go files in $GOPATH/src/github.com/ethanpil/sheret`, don't bother it.
2. Go to the $GOPATH/src/github.com/ethanpil/sheret/src.
3. Download libraries with `go get`.
4. Compile `go build -ldflags="-s -w" -o sheret sheret.go`

**Comment by rstaub77** [comment]
Thanks for your answer.
It's a little pity though, as the readme says "There are no dependencies.".
I suspect it is more a packaging issue.

**Comment by ethanpil** [comment]
Good call. Since I added the open browser flag, there is now a dependency on the Browser package for cross platform browser open. I will investigate alternative. Is the zero dependency a real issue for you? Why?

**Comment by rstaub77** [comment]
Hi,
The zero dependency is not such an issue, except that documentation advertises it.
One of the quality of Go is its portability, and as no release executable is provided for Linux, I tried to compile the sources myself to get a monolithic executable to work on my servers.
Usually a simple "go build" in the source directory is enough to get the compilation done.
In this specific case, it crashes, and I don't know Go this well to know for sure what is going wrong.
I think a little update of the documentation and a proper paragraph about compiling the project would be enough.

**Comment by pixelblender** [comment]
same issue on macos big sur: `sheret.go:20:3: no required module provides package github.com/pkg/browser: go.mod file not found in current directory or any parent directory; see 'go help modules'` how to fix the issue?

**Comment by ethanpil** [comment]
Thanks for letting me know. I will take a look when I have some time.

**Comment by saukrs** [comment]
Is it fixed yet?

**Comment by ethanpil** [comment]
I think I will remove the Browser functionality and package dependency and release a new version.

**Comment by elmkni** [comment]
Hi, my name is Elmar and I think I found a solution to the problem ...

[Go](https://go.dev/) 1.20.5 has browser access build in: `go\src\cmd\internal\browser\browser.go` !

Sadly internal code is not allowed to be imported directly, but of course one can copy the contained functions to 'sheret.go':
```go
import (
	"os/exec"
	"runtime"
	"time"
)

// Commands returns a list of possible commands to use to open a url.
func Commands() [][]string {
	var cmds [][]string
	if exe := os.Getenv("BROWSER"); exe != "" {
		cmds = append(cmds, []string{exe})
	}
	switch runtime.GOOS {
	case "darwin":
		cmds = append(cmds, []string{"/usr/bin/open"})
	case "windows":
		cmds = append(cmds, []string{"cmd", "/c", "start"})
	default:
		if os.Getenv("DISPLAY") != "" {
			// xdg-open is only for use in a desktop environment.
			cmds = append(cmds, []string{"xdg-open"})
		}
	}
	cmds = append(cmds,
		[]string{"chrome"},
		[]string{"google-chrome"},
		[]string{"chromium"},
		[]string{"firefox"},
	)
	return cmds
}

// Open tries to open url in a browser and reports whether it succeeded.
func Open(url string) bool {
	for _, args := range Commands() {
		cmd := exec.Command(args[0], append(args[1:], url)...)
		if cmd.Start() == nil && appearsSuccessful(cmd, 3*time.Second) {
			return true
		}
	}
	return false
}

// appearsSuccessful reports whether the command appears to have run successfully.
// If the command runs longer than the timeout, it's deemed successful.
// If the command runs within the timeout, it's deemed successful if it exited cleanly.
func appearsSuccessful(cmd *exec.Cmd, timeout time.Duration) bool {
	errc := make(chan error, 1)
	go func() {
		errc <- cmd.Wait()
	}()

	select {
	case <-time.After(timeout):
		return true
	case err := <-errc:
		return err == nil
	}
}
```
delete `"github.com/pkg/browser"` from the import section of 'sheret.go' and change the line:

`browser.OpenURL("http://localhost:"+*port)`

to:

`Open("http://localhost:"+*port)`

and 'sheret.go' is compilable again without the hassle of go.mod/go.sum, finally without external dependencies and even 13 KB smaller than before! Btw: [Go](https://go.dev/) 1.20.5 already produces smaller executables than older [Go](https://go.dev/) versions (at least on Windows.)!

### So there is no need to 'remove the Browser functionality' ...

I hope this helped to keep sheret as it is: **a wonderful, useful and ingenious tool for serving directories/files** etc.
I use it a lot in conjunction with [x_ite](https://create3000.github.io/x_ite) (a nice browser-based X3D viewer).

Regards,

Elmar

**Comment by elmkni** [comment]
Hi, it's Elmar again ...

If you just want to make 'sheret.go' compilable **without** changing the code, you could alternatively add a 'go.mod' and 'go.sum' to the directory where 'sheret.go' resides.

go.mod:
```
module github.com/ethanpil/sheret

go 1.20

require	github.com/pkg/browser v0.0.0-20210911075715-681adbf594b8
require golang.org/x/sys       v0.0.0-20210616045830-e2b7044e8c71 // indirect
```

go.sum:
```
github.com/pkg/browser v0.0.0-20210911075715-681adbf594b8 h1:KoWmjvw+nsYOo29YJK9vDA65RGE3NrOnUtO7a+RF9HU=
github.com/pkg/browser v0.0.0-20210911075715-681adbf594b8/go.mod h1:HKlIX3XHQyzLZPlr7++PzdhaXEj94dEiJgZDTsxEqUI=
golang.org/x/sys v0.0.0-20210616045830-e2b7044e8c71 h1:X/2sJAybVknnUnV7AD2HdT6rm2p5BP6eH2j+igduWgk=
golang.org/x/sys v0.0.0-20210616045830-e2b7044e8c71/go.mod h1:oPkhp1MJrh7nUepCBck5+mAzfO9JrbApNNgaTdGDITg=
```
and now just: 

```
go build -ldflags="-s -w" sheret.go
strip -s sheret.exe
```
with best regards,

Elmar

**Comment by ethanpil** [comment]
Hi @elmkni thanks for the updates. If you submit a PR I will merge it and release a new version.

**Comment by elmkni** [comment]
Hi Ethan,

I've just submitted a pull request.

But I made a mistake and added 'go.mod' and 'go.sum' to the 'src' directory.

## So do not merge my fork!

Instead you should create the missing files by yourself with just two commands:

`go mod init sheret`

and

`go mod tidy`

from within your root directory, where 'go.mod' and 'go.sum' will be created!

The first command creates a suitable 'go.mod' file, the second command updates the needed dependencies and creates 'go.sum'.

Now executing 'build.bat' from the 'build' directory will produce 'sheret.exe' without problems!

With best regards,

Elmar

**Comment by elmkni** [comment]
Hi Ethan,

**_I've corrected my fork/pull-request_** by moving 'go.mod' and 'go.sum' from the 'src' directory to the parent directory, where they belong!

So now you can **_either_** merge my fork **_or_** do `go mod init sheret` / `go mod tidy` yourself, whatever suits you best ..

With best regards,

Elmar

**Comment by elmkni** [comment]
Hi Ethan,

I've changed the dependencies from "github.com/pkg/browser" to "github.com/icza/gox/osx" because it is much smaller and has no sub-dependency itself! I also added the missing licenses for Go and Gox (they are actually compulsory) and edited 'release.bat' accordingly. I also bumped the version from '1.2' to '1.22' to reflect the changes I made! I've also added the go-1.20.5 compiled sheret binary v1.22 executable and release archive!

With best regards,

Elmar
